src/resolve/resolved-link.c

```diff
@@ -528,6 +524,10 @@ static void link_read_settings(Link *l) {
         r = link_update_search_domains(l);
         if (r < 0)
                 log_warning_errno(r, "Failed to read search domains for interface %s, ignoring: %m", l->name);
+
+        r = link_update_dns_servers(l);
+        if (r < 0)
+                log_warning_errno(r, "Failed to read DNS servers for interface %s, ignoring: %m", l->name);
```

Comment:
not sure I follow, can you elaborate? why precisely does the order matter here?

Reply:
The purpose is that link_read_settings() is meant to read in the configurable items for a link from the configuration files such as /etc/systemd/network/*.conf and then define a DNS server on the basis of those settings.

The code that does that setup is link_update_dns_servers() by eventually calling dns_server_new(). however in the current version of systemd that operation is done before all settings have been read in and recorded against the "link struct".

When I made my change to make the best feature level configurable (in another pull request) I found that the DNS object was not seeing the setting come through. In the future if further configurability of systemd-resolved is made then they too will come up against this problem if there is some data item from the link that should be propagated to the DNS server.

Reply:
still not following. Why does it matter whether the `Link` struct is fully up-to-date before dns_server_new() is called? that function doesn't really copy anything out from the link object, so why does it matter?

Reply:
Ok, so don't pass in the link struct to the dns_server_new() function. Since it will never be consulted. That will avoid all problems in the future. There are items that are hardcoded in there, which some RFEs ask for to become configurable and I thought that meant that it would come from the link and propagate to the DNS server, but I was obviously mistaken?

Reply:
We manage DNS servers by link, and we pass the Link object into dns_server_new() to indicate this relationship, so that the DnsServer object takes a reference to the Link object, but this remains a reference and no data is copied between the two objects. Or to say this differently: whenever we need to figure out information about a link a DnsServer is attached to we follow the reference to the orginal object, and thus take into account whatever the current properties of the Link are.

Hence I don#t really see what this patch really is supposed to do...

Reply:
Maybe it is my fault, but I thought this was a separate item to #7002 so I broke it out. The real context is in #7002 because in that there is a use case where it becomes relevant.

But to avoid jumbling the two conversations up, my view point is simply this:

1)

A DNS server is a result of reading in a configuration of a link. Does it make sense to create the DNS server before the configuration of the link has been read in?

It's akin to leaving variables uninitialised. You are saying that it doesn't matter because the current code in the "DNS object" is not using any more data from the "Link object" and therefore the patch is pointless.

In #7002 I spent several hours scratching my head trying to understand why addition to the  configurability of a link didn't come through.

2)

Does it make sense to always keep the details of the DNS object within the link object? In the case where you changed the DNS object to handle feature levels then you do that by DNS object and not Link object, consequently in #7002 I followed the same pattern and then it became a problem that the Link object wasn't finished yet when I came to copy across the best possible feature level.